specs/dotsigstore_bundle/pb/bundle.proto

```diff
@@ -0,0 +1,84 @@
+syntax = "proto3";
+package sigstore;
+
+// https://raw.githubusercontent.com/secure-systems-lab/dsse/9c813476bd36de70a5738c72e784f123ecea16af/envelope.proto
+import "envelope.proto";
+
+option go_package = "github.com/sigstore/xxx/bundle";
+
+enum HashAlgorithm {
+        SHA2_224 = 0;
+        SHA2_256 = 1;
+        SHA2_384 = 2;
+        SHA2_512 = 3;
+        SHA2_512_224 = 4;
+        SHA2_512_256 = 5;
+}
+
+message AttestationBlob {
+        // In JSON encoding this shall be base64 encoded.
+        bytes payload_hash = 1;
+        HashAlgorithm algorithm = 2;
+        // In JSON encoding this shall be base64 encoded.
+        bytes signature = 3;
+}
+
+// RekorEntry is a separate type, capturing all the details required
+// from Rekor to reconstruct an entry, given that the payload is provided
+// via other means.
+// This type can easily be created from the existing response from Rekor.
+// Future iterations could rely on Rekor returning the minimal set of
+// attributes (excluding the payload) that are required for verifying the
+// SET.
+message RekorEntry {
+        int64 log_index = 1;
+        string log_id = 2;
+        // Even if kind can be derived from the attestation type it's kept
+        // as this struct could be automatically created by the sigstore
+        // library based on the response from Rekor. Also during verification
+        // this struct can be passed to the library for Rekor entry
+        // reconstruction. By duplicating the kind here we would so minimize
+        // any possibility of error due to attestation/kind matching logic,
+        // as that logic must be hard-coded in the client. With this construct
+        // it's only determined during publish to Rekor.
+        string kind = 3;
+        string version = 4;
+        string signed_entry_timestamp = 5;
+        // Note that this will be encoded as a string in JSON:
+        // https://developers.google.com/protocol-buffers/docs/proto3#json
+        int64 integrated_time = 6;
+}
+
+message PublicKey {
+        // Optional unauthenticated hint on which key to use.
+        // If a DSSE envelope is signed, at least one of the signatures
+        // MUST have a signature with the same key id.
+        string key_id = 1;
+}
+
+message X509Cert {
+        // PEM encoded certificate.
+        string certificate = 1;
+        // PEM encoded certificate chain.
+        string chain = 2;
+}
+
+message Bundle {
+        // MUST be application/vnd.dev.sigstore.bundle.v1+json
+        string media_type = 1;
+        optional RekorEntry rekor_entry = 2;
```

Comment:
The timestamp can be provided in addition to the rekor entry. There's two ways we use rekor - as a proof of inclusion in a log, and as a countersignature over a timestamp. Timestamping won't provide the former, only the second. 

I don't think we want it as a oneof, because we might provide both a user-provided timestamp and a rekor entry.

@asraa

Reply:
Even if Rekor returns a timestamp, I still think this should live outside of the `RekorEntry` message. It's just an optimization if Rekor returns a timestamp - the timestamp will still come from a separate service in Sigstore.

Reply:
Makes sense, I think we'd just have to think about naming.

In an ideal state, we have a set of signed timestamps (3161 or roughtime, from both Sigstore and user-provided), and an inclusion proof from the log. Bundling them in the same message is reasonable because we need both to trust an entry. So maybe the options are:

* SET - Provides timestamp and promise of inclusion
* SET + signed timestamps - Provides timestamps from both sigstore and externally provided, and a promise of inclusion
* Signed timestamps + Inclusion proof - Provides timestamp and proof(-ish, minus consistency) of inclusion

What I really want is a oneof around (SET, signed timestamp) and a oneof (SET, inclusion proof), but that duplication isn't ideal. Maybe we just group them all together?

Reply:
> What I really want is a oneof around (SET, signed timestamp) and a oneof (SET, inclusion proof), but that duplication isn't ideal. Maybe we just group them all together?

It would be really nice from a type system level to have a `oneof` where *any* of the possible values means that the signature is trusted (as long as the values themselves are valid).

IMO the duplication is worth it, because it simplifies client code.

Reply:
Here's an example:

```
message VerificationStuff { // name TBD
  one_of timestamp {
    string SET_timestamp; // use only for timestamping
    repeated string signed_timestamps; // a set of user and Sigstore provided timestamps 
  }

  one_of proof {
    string SET; // use only for checking inclusion
    InclusionProof p; // contains inclusion proof and STH
  }
}
```

The only question is what happens if the SETs mismatch for some reason? Are we ok with just making clients check that they match, because we know we'd like to eventually deprecate them?

Reply:
Hm....looking it it, that's a little unfortunate. Maybe something like:

```proto
message VerificationStuff { // name TBD
  string SET;

  one_of timestamp {
    bool use_set;  // use only for timestamping
    repeated string signed_timestamps; // a set of user and Sigstore provided timestamps 
  }

  one_of proof {
    bool use_set; // use only for checking inclusion
    InclusionProof p; // contains inclusion proof and STH
  }
}
```

Reply:
I like that!

fyi @asraa

Reply:
Updating the proposal to reflect this.

Reply:
Taking a second thought around this, my understanding is that the example does not include the case where only RFC3161 timestamps are included (unless `SET` is empty and so treated as an optional parameter). Adding that and making sure the protobuf file compiles properly (repeated use of `use_set` in different `oneof`s is prohibited), my proposal would be this. I would think this is similar to the above with a combination of two `oneof`s, but they are folded out to the the different possible combinations.

```protobuf
// A list of RFC3161 signed timestamps.
message SignedTimestamps {
        repeated string signed_timestamps = 1;
}

// This message contains the transparency log entry (inclusion promise) and
// signed timestamps, provided by sigstore and/or the user.
message TimestampedTransparencyLogEntry {
        TransparencyLogEntry tlog_entry = 1;
        SignedTimestamps timestamps = 2;
}

// This message contains the transparency log's proof of inclusion and
// signed timestamps, provided by sigstore and/or the user.
message TimestampedInclusionProof {
        InclusionProof inclusion_proof = 1;
        SignedTimestamps timestamps = 2;
}

message VerificationData {
        oneof data {
                // This is the regular inclusion promise and timestamp from
                // the transparency log
                TransparencyLogEntry tlog_entry = 1;
                // This is the regular inclusion promise and timestamp from
                // the transparency log, and extra RFC3161 signed timestamps
                // provided by Sigstore and user.
                TimestampedTransparencyLogEntry timestamped_tlog_entry = 2;
                // The inclusion proof from the transparency log, and RFC3161
                // timestamps provided by Sigstore and the user.
                TimestampedInclusionProof timestamped_inclusion_proof = 3;
                // A list of RFC3161 signed timestamps provided by the user.
                SignedTimestamps timestamps = 4;
        }
}

```

Reply:
Love this design! I think this is an improvement on an empty `string SET`. This looks good, a couple clarifying questions:

To confirm, we want to support the case of providing only signed timestamps, correct? This would be in the case where you don't want to upload to Rekor for some reason (let's say a private package). I think yes, but just wanted to check that we're intentional about `SignedTimestamps timestamps = 4;`. Maybe a comment about this?

I left a comment previously about including both the inclusion proof and promise vs one or the other. What do you think? I'm thinking that until we know exactly what we want to do with the inclusion proof, I think it would be good to include both. So the 4 cases are then:

1. A promise + inclusion proof, and timestamp comes from promise
2. A promise + inclusion proof, and timestamp comes from signed timestamp
3. An inclusion proof + timestamp, and timestamp comes from signed timestamp
4. Just a timestamp

One question that comes up is if `InclusionProof` is included in `TransparencyLogEntry`, then do we need (2) and (3) being separate?

Reply:
> To confirm, we want to support the case of providing only signed timestamps, correct? 

Yes, I'll clarify that.

> I left a comment previously about including both the inclusion proof and promise vs one or the other.

Yeah, this requires a bit more thinking. My first draft (before thinking more about _this_ discussion) was to include the inclusion proof in the `TransparencyLogEntry`. And I'm in favour of simplifying the data structure, so I would prefer to include the proof, as this will get us three different options, and then the verifier can decide if they want to validate the inclusion proof or not.

Reply:
Can we resolve this discussion?